README.md
Upton
==========
Upton is a framework for easy web-scraping with a useful debug mode that doesn't hammer your target's servers. It does the repetitive parts of writing scrapers, so you only have to write the unique parts for each site.

Documentation
----------------------

With Upton, you can scrape complex sites to a CSV in just a few lines of code:

```ruby
scraper = Upton::Scraper.index("http://www.propublica.org", "section#river h1 a")
scraper.scrape_to_csv "output.csv" do |html|
  Nokogiri::HTML(html).search("#comments h2.title-link").map &:text
end
```

Just specify a URL to a list of links (an "index"), an XPath expression or CSS selector for the links and a block of what to do with the content of the pages you've scraped. Upton comes with some pre-written blocks (Procs, technically) for scraping simple lists and tables, like the `list` function above.

Upton operates on the theory that, for most scraping projects, you need to scrape two types of pages:

1. Instance pages, which are the goal of your scraping, e.g. job listings or news articles.
2. Index pages, which list instance pages. For example, a job search site's search page or a newspaper's homepage.

You can also directly specify the list of instance page URLs to scrape:

```ruby
scraper = Upton::Scraper.instances(["http://www.propublica.org/article_1.html", "http://www.propublica.org/article_2.html"])
scraper.scrape_to_csv "output.csv" do |html|
  Nokogiri::HTML(html).search("#comments h2.title-link").map &:text
end
```

For more complex use cases, subclass `Upton::Scraper` and override the relevant methods. If you're scraping links from an API, you would override `get_index`; if you need to log in before scraping a site or do something special with the scraped instance page, you would override `get_instance`.

The `get_instance` and `get_index` methods use a protected method `get_page(url)` which, well, gets a page. That's not very special. The more interesting part is that `get_page(url, stash)` transparently stashes the response of each request if the second parameter, `stash`, is true. Whenever you repeat a request (with `true` as the second parameter), the stashed HTML is returned without going to the server. This is helpful in the development stages of a project when you're testing some aspect of the code and don't want to hit a server each time. If you are using `get_instance` and `get_index`, this can be en/disabled per instance of `Upton::Scraper` or its subclasses with the `@debug` option. Setting the `stash` parameter of the `get_page` method should only be used if you've overridden `get_instance` or `get_index` in a subclass.

Upton also sleeps (by default) 30 seconds between non-stashed requests, to reduce load on the server you're scraping. This is configurable with the `@sleep_time_between_requests` option.

Upton can handle pagination too. Scraping paginated index pages that use a query string parameter to track the current page (e.g. `/search?q=test&page=2`) is possible by setting `@paginated` to true. Use `@pagination_param` to set the query string parameter used to specify the current page (the default value is `page`). Use `@pagination_max_pages` to specify the number of pages to scrape (the default is two pages). You can also set @pagination_interval` if you want to increment pages by a number other than 1 (i.e. if the first page is 1 and lists instances 1 through 20, the second page is 21 and lists instances 21-41, etc.) See the Examples section below.

To handle non-standard pagination, you can override the `next_index_page_url` and `next_instance_page_url` methods; Upton will get each page's URL returned by these functions and return their contents.

<b>For more complete documentation</b>, see [the RDoc](http://rubydoc.info/gems/upton/frames/index).

<b>Important Note:</b> Upton is alpha software. The API may change at any time. 

####How is this different than Nokogiri?
Upton is, in essence, sugar around RestClient and Nokogiri. If you just used those tools by themselves to write scrapers, you'd be responsible for writing code to fetch, save (maybe), debug and sew together all the pieces in a slightly different way for each scraper. Upton does most of that work for you, so you can skip the boilerplate.

####Upton doesn't quite fit your needs?
Here are some similar libraries to check out for inspiration. No promises, since I've never used them, but they seem similar and were [recommended by various HN commenters](https://news.ycombinator.com/item?id=6086031): 

- [Pismo](https://github.com/peterc/pismo)
- [Spidey](https://github.com/joeyAghion/spidey)
- [Anemone](http://anemone.rubyforge.org/)
- [Pupa.rb](https://github.com/opennorth/pupa-ruby) / [Pupa](https://github.com/opencivicdata/pupa)

And these are some libraries that do related things:

- [SelectorGadget](http://selectorgadget.com/)
- [HayStax](https://github.com/danhillreports/haystax)


Examples
----------------------
If you want to scrape ProPublica's website with Upton, this is how you'd do it. (Scraping our [RSS feed](http://feeds.propublica.org/propublica/main) would be smarter, but not every site has a full-text RSS feed...)

```ruby
scraper = Upton::Scraper.new("http://www.propublica.org", "section#river section h1 a")
scraper.scrape do |article_html_string|
  puts "here is the full html content of the ProPublica article listed on the homepage: "
  puts "#{article_html_string}"
  #or, do other stuff here.
end
```

Simple sites can be scraped with pre-written `list` block in `Upton::Utils', as below:

```ruby
scraper = Upton::Scraper.new("http://nytimes.com", "ul.headlinesOnly a")
scraper.scrape_to_csv("output.csv", &Upton::Utils.list("h6.byline"))
```

A `table` block also exists in `Upton::Utils` to scrape tables to an array of arrays, as below:

```ruby
> scraper = Upton::Scraper.new(["http://website.com/story.html"])
> scraper.scrape(&Upton::Utils.table("//table[2]"))
[["Jeremy", "$8.00"], ["John Doe", "$15.00"]]
```

This example shows how to scrape the first three pages of ProPublica's search results for the term `tools`:

```ruby
> scraper = Upton::Scraper.new("http://www.propublica.org/search/search.php?q=tools",
                               ".compact-list a.title-link")
> scraper.paginated = true
> scraper.pagination_param = 'p'    # default is 'page'
> scraper.pagination_max_pages = 3  # default is 2
> scraper.scrape_to_csv("output.csv", &Upton::Utils.list("h2"))
```


Contributing
----------------------
I'd love to hear from you if you're using Upton. I also appreciate your suggestions/complaints/bug reports/pull requests. If you're interested, check out the issues tab or [drop me a note](http://github.com/jeremybmerrill).

In particular, if you have a common, *abstract* use case, please add them to [lib/utils.rb](https://github.com/propublica/upton/blob/master/lib/utils.rb). Check out the `table_to_csv` and `list_to_csv` methods for examples.

(The pull request process is pretty easy. Fork the project in Github (or via the `git` CLI), make your changes, then submit a pull request on Github.) 

Why "Upton"
----------------------
Upton Sinclair was a pioneering, muckraking journalist who is most famous for _The Jungle_, a novel portraying the reality of immigrant labor struggles in Chicago meatpacking plants at the start of the 1900s. Upton, the gem, sprang out of a ProPublica project pertaining to labor issues.

Notes
------------------------
Test data is copyrighted by either ProPublica or various Wikipedia contributors.
In either case, it's reproduced here under a Creative Commons license. In ProPublica's case, it's BY-NC-ND; in Wikipedia's it's BY-SA.


changelog
2014-02-16  Jeremy B. Merrill  <jeremy.merrill@propublica.org>

  * adds HTML comment to stashed pages with date/time downloaded, URL
  * @pagination_interval option
  * separate Upton::Scraper into its own file

2014-02-15  Jeremy B. Merrill  <jeremy.merrill@propublica.org>
  * Various other changes. Sorry, forgot to keep track... :/

2013-05-30  Jeremy B. Merrill  <jeremy.merrill@propublica.org>

  * upton.gemspec: initial release
  * upton.rb: initial release

spec/upton_spec.rb
# encoding: UTF-8

require 'rack'
require 'thin'
require 'nokogiri'
require 'restclient'
require 'fileutils'
require "spec_helper.rb"

require './lib/upton'


describe Upton do
  before :all do
    @headlines = ["Webinar: How to Use Prescriber Checkup to Power Your Reporting",
                 "",
                 "A Prosecutor, a Wrongful Conviction and a Question of Justice",
                 "Six Facts Lost in the IRS Scandal"]
    @most_commented_heds = [["Six Facts Lost in the IRS Scandal",
                        "How the IRS’s Nonprofit Division Got So Dysfunctional",
                        "Sound, Fury and the IRS Mess",
                        "The Most Important #Muckreads on Rape in the Military",
                        "Congressmen to Hagel: Where Are the Missing War Records?",
                        "As Need for New Flood Maps Rises, Congress and Obama Cut Funding",
                        "A Prosecutor, a Wrongful Conviction and a Question of Justice",
                        "A Prolonged Stay: The Reasons Behind the Slow Pace of Executions",
                        "The Story Behind Our Hospital Interactive",
                        "irs-test-charts-for-embedding"]]
    @east_timor_prime_ministers = [[
                                    ["#", "Portrait", "Name(Birth–Death)", "Term of Office", "Party",
                                      "1", "2", "3", "4",],
                                    [],
                                    ["", "Mari Alkatiri(b. 1949)", "20 May 2002", "26 June 2006[1]", "FRETILIN"],
                                    ["", "José Ramos-Horta(b. 1949)", "26 June 2006", "19 May 2007", "Independent"],
                                    ["", "Estanislau da Silva(b. 1952)", "19 May 2007", "8 August 2007", "FRETILIN"],
                                    ["", "Xanana Gusmão(b. 1946)", "8 August 2007", "Incumbent", "CNRT"],
                                  ]]
    @searchResults = ["Webinar: How to Use Prescriber Checkup to Power Your Reporting",
                 "A Prosecutor, a Wrongful Conviction and a Question of Justice",
                 "Six Facts Lost in the IRS Scandal"]
  end

  it "should scrape in the basic case" do
    stub_request(:get, "www.example.com/propublica.html").
      to_return(:body => File.new('./spec/data/propublica.html'), :status => 200)
    stub_request(:get, "www.example.com/discussion.html").
      to_return(:body => File.new('./spec/data/discussion.html'), :status => 200)
    stub_request(:get, "www.example.com/prosecutor.html").
      to_return(:body => File.new('./spec/data/prosecutor.html'), :status => 200)
    stub_request(:get, "www.example.com/webinar.html").
      to_return(:body => File.new('./spec/data/webinar.html'), :status => 200)
    stub_request(:get, "www.example.com/sixfacts.html").
      to_return(:body => File.new('./spec/data/sixfacts.html'), :status => 200)

    propubscraper = Upton::Scraper.index("http://www.example.com/propublica.html", "section#river section h1 a")
    propubscraper.debug = true
    propubscraper.verbose = false
    propubscraper.sleep_time_between_requests = 0
    propubscraper.stash_folder = "test_stashes"

    heds = propubscraper.scrape do |doc|
      doc.css('h1.article-title').text
    end
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    heds.should eql @headlines
  end

  it 'should properly handle relative urls'  do
# uses a modified page from the previous test in which the target
# href, http://127.0.0.1:9876/prosecutors.html, has been changed
# to a relative url
#
# Note: this test is a bit quirky, because it passes on the fact that
# the resolve_url creates a url identical to one that is already stashed ("prosecutors.html").
# So it works, but because of a coupling to how Upton handles caching in the file system

    stub_request(:get, "www.example.com/propublica-relative.html").
      to_return(:body => File.new('./spec/data/propublica-relative.html'), :status => 200)
    stub_request(:get, "www.example.com/prosecutor.html").
      to_return(:body => File.new('./spec/data/prosecutor.html'), :status => 200)
    stub_request(:get, "www.example.com/sixfacts.html").
      to_return(:body => File.new('./spec/data/sixfacts.html'), :status => 200)
    stub_request(:get, "www.example.com/webinar.html").
      to_return(:body => File.new('./spec/data/webinar.html'), :status => 200)
    stub_request(:get, "www.example.com/discussion.html").
      to_return(:body => File.new('./spec/data/discussion.html'), :status => 200)


    propubscraper = Upton::Scraper.index("http://www.example.com/propublica-relative.html", "section#river h1 a")
    propubscraper.debug = true
    propubscraper.verbose = false
    propubscraper.sleep_time_between_requests = 0
    propubscraper.stash_folder = "test_stashes"

    heds = propubscraper.scrape do |doc|
      hed = doc.css('h1.article-title').text
    end
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    heds.should eql ["A Prosecutor, a Wrongful Conviction and a Question of Justice"]
  end

  it "should scrape a list properly with the list helper" do
    stub_request(:get, "www.example.com/propublica.html").
      to_return(:body => File.new('./spec/data/propublica.html'), :status => 200)

    propubscraper = Upton::Scraper.instances(["http://www.example.com/propublica.html"])
    propubscraper.debug = true
    propubscraper.verbose = false
    propubscraper.sleep_time_between_requests = 0
    propubscraper.stash_folder = "test_stashes"

    list = propubscraper.scrape(&Upton::Utils.list("#jamb.wNarrow #most-commented li a"))
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    list.should eql @most_commented_heds
  end

  it "should scrape a table properly with the table helper" do
    stub_request(:get, "www.example.com/easttimor.html").
      to_return(:body => File.new('./spec/data/easttimor.html'), :status => 200)

    propubscraper = Upton::Scraper.instances(["http://www.example.com/easttimor.html"])
    propubscraper.debug = true
    propubscraper.verbose = false
    propubscraper.sleep_time_between_requests = 0
    propubscraper.stash_folder = "test_stashes"

    table = propubscraper.scrape(&Upton::Utils.table('//table[contains(concat(" ", normalize-space(@class), " "), " wikitable ")][2]'))
    table.map{|outer| outer.map{|row| row.map{|cell| cell.gsub!("\n", '') } }} # cope with diff nokogiri versions differing behavior.
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    table.should eql @east_timor_prime_ministers
  end

  it "should test saving files with the right encoding" do
    pending "finding a site that gives funny encodings"
  end

  it "should scrape paginated pages" do
    stub_request(:get, "www.example.com/propublica_search.html").
      to_return(:body => File.new('./spec/data/propublica_search.html'), :status => 200)
    stub_request(:get, "www.example.com/propublica_search.html?p=2").
      to_return(:body => File.new('./spec/data/propublica_search_page_2.html'), :status => 200)
    stub_request(:get, "www.example.com/propublica_search.html?p=3").
      to_return(:body => '', :status => 200)
    stub_request(:get, "www.example.com/webinar.html").
      to_return(:body => File.new('./spec/data/webinar.html'), :status => 200)
    stub_request(:get, "www.example.com/prosecutor.html").
      to_return(:body => File.new('./spec/data/prosecutor.html'), :status => 200)
    stub_request(:get, "www.example.com/sixfacts.html").
      to_return(:body => File.new('./spec/data/sixfacts.html'), :status => 200)


    propubscraper = Upton::Scraper.index(
        "http://www.example.com/propublica_search.html", 
        '.compact-list a.title-link',
        {
          :paginated => true,
          :pagination_param => 'p',
          :pagination_max_pages => 3,
        }
      )
    propubscraper.debug = true
    propubscraper.verbose = false
    propubscraper.sleep_time_between_requests = 0
    propubscraper.stash_folder = "test_stashes"

    results = propubscraper.scrape do |doc|
      hed = doc.css('h1.article-title').text
    end
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    results.should eql @searchResults
  end


  before do
    Upton::Scraper.stub(:sleep)
  end

  it "should sleep after requests with caching disabled" do
    stub_request(:get, "www.example.com")
    u = Upton::Scraper.index("http://www.example.com", '.whatever')
    u.index_debug = false
    u.sleep_time_between_requests = 1 #don't sleep too long, that's annoying.
    u.should_receive(:sleep)
    u.scrape
  end

  it "should sleep after uncached requests when caching is enabled" do
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    stub_request(:get, "www.example.com")
    u = Upton::Scraper.index("http://www.example.com", '.whatever')
    u.index_debug = true
    u.stash_folder = "test_stashes"
    u.sleep_time_between_requests = 1 #don't sleep too long, that's annoying.
    u.should_receive(:sleep)
    u.scrape
  end

  it "should sleep after paginated requests when caching is disabled" do
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    stub_request(:get, "www.example.com/propublica_search.html").
      to_return(:body => File.new('./spec/data/propublica_search.html'), :status => 200)
    stub_request(:get, "www.example.com/propublica_search.html?p=2").
      to_return(:body => File.new('./spec/data/propublica_search_page_2.html'), :status => 200)
    stub_request(:get, "www.example.com/propublica_search.html?p=3").
      to_return(:body => '', :status => 200)
    stub_request(:get, "www.example.com/webinar.html").
      to_return(:body => File.new('./spec/data/webinar.html'), :status => 200)
    stub_request(:get, "www.example.com/prosecutor.html").
      to_return(:body => File.new('./spec/data/prosecutor.html'), :status => 200)
    stub_request(:get, "www.example.com/sixfacts.html").
      to_return(:body => File.new('./spec/data/sixfacts.html'), :status => 200)


    u = Upton::Scraper.index("http://www.example.com/propublica_search.html", '.nonexistent',
        {
          :paginated => true,
          :pagination_param => 'p',
          :pagination_max_pages => 3,
        }
      )
    u.index_debug = false
    u.debug = false
    u.sleep_time_between_requests = 1 #don't sleep too long, that's annoying.
    u.stash_folder = "test_stashes"

    u.should_receive(:sleep).exactly(3).times #once for each search page, so 3.
    u.scrape
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
  end


  it "should save to the designated stash folder" do
    custom_cache_folder = "#{Dir.tmpdir}/upton/test"
    FileUtils.rm_rf(custom_cache_folder)
    stub_request(:get, "www.example.com").
      to_return(:body => '', :status => 200)

    u = Upton::Scraper.index("http://www.example.com", '.whatever')
    u.sleep_time_between_requests = 0.0
    u.stash_folder = custom_cache_folder
    u.debug = true
    u.scrape do
      1+1
    end
    files = Dir.glob(custom_cache_folder)
    expect(files).not_to be_empty
  end

  it "should scrape in the basic case with the index method" do
    stub_request(:get, "www.example.com/propublica.html").
      to_return(:body => File.new('./spec/data/propublica.html'), :status => 200)
    stub_request(:get, "www.example.com/discussion.html").
      to_return(:body => File.new('./spec/data/discussion.html'), :status => 200)
    stub_request(:get, "www.example.com/prosecutor.html").
      to_return(:body => File.new('./spec/data/prosecutor.html'), :status => 200)
    stub_request(:get, "www.example.com/webinar.html").
      to_return(:body => File.new('./spec/data/webinar.html'), :status => 200)
    stub_request(:get, "www.example.com/sixfacts.html").
      to_return(:body => File.new('./spec/data/sixfacts.html'), :status => 200)

    propubscraper = Upton::Scraper.index("http://www.example.com/propublica.html", "section#river section h1 a")
    propubscraper.debug = true
    propubscraper.verbose = false
    propubscraper.sleep_time_between_requests = 0
    propubscraper.stash_folder = "test_stashes"

    heds = propubscraper.scrape do |doc|
      hed = doc.css('h1.article-title').text
    end
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    heds.should eql @headlines
  end

  it "should allow instances to be set on a new Scraper" do
    stub_request(:get, "www.example.com/propublica.html").
      to_return(:body => File.new('./spec/data/propublica.html'), :status => 200)
    stub_request(:get, "www.example.com/discussion.html").
      to_return(:body => File.new('./spec/data/discussion.html'), :status => 200)
    stub_request(:get, "www.example.com/prosecutor.html").
      to_return(:body => File.new('./spec/data/prosecutor.html'), :status => 200)
    stub_request(:get, "www.example.com/webinar.html").
      to_return(:body => File.new('./spec/data/webinar.html'), :status => 200)
    stub_request(:get, "www.example.com/sixfacts.html").
      to_return(:body => File.new('./spec/data/sixfacts.html'), :status => 200)

    propubscraper = Upton::Scraper.instances(["www.example.com/webinar.html",
                                              "www.example.com/discussion.html", 
                                              "www.example.com/prosecutor.html", 
                                              "www.example.com/sixfacts.html"])

    propubscraper.debug = true
    propubscraper.verbose = false
    propubscraper.sleep_time_between_requests = 0
    propubscraper.stash_folder = "test_stashes"

    heds = propubscraper.scrape do |doc|
      hed = doc.css('h1.article-title').text
    end
    FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
    heds.should eql @headlines
  end

  it "should allow Scrapers to be added (indexes)" do
    u = Upton::Scraper.index("http://www.example1.com", '.link')
    w = Upton::Scraper.index("http://www.example2.com", '.link')
    new_scraper = u + w
    new_scraper.instance_variable_get(:@indexes).map{|a| a[0]}.should eql ["http://www.example1.com", "http://www.example2.com"]
  end

  it "should allow Scrapers to be added (instances)" do
    pending
    u = Upton::Scraper.instances(["http://www.example1.com"])
    w = Upton::Scraper.instances(["http://www.example2.com"])
    new_scraper = u + w
    new_scraper.instance_variable_get(:@indexes).should eql []
    new_scraper.instance_variable_get(:@instance_urls).map{|a| a[0]}.should eql ["http://www.example1.com", "http://www.example2.com"]
  end


  before do
    Upton::Scraper.stub(:puts)
  end

  it "should be silent if verbose is false" do
    stub_request(:get, "www.example.com")
    u = Upton::Scraper.index("http://www.example.com", '.whatever')
    u.sleep_time_between_requests = 0.0
    u.verbose = false
    u.should_not_receive(:puts)
    u.scrape
  end

end


test/test_upton.rb
# encoding: UTF-8

require 'test/unit'
require 'rack'
require 'thin'
require 'nokogiri'
require 'restclient'
require './lib/upton'
require 'fileutils'

module Upton
  module Test

    # class ProPublicaScraper < Upton::Scraper
    #   def initialize(a, b, c)
    #     super
    #     @verbose = false
    #     @debug = false
    #     @stash_folder = "test_stashes"
    #   end
    # end


    class UptonTest < ::Test::Unit::TestCase

      # def test_get_page
      #TODO
      # end

      # def test_stash
      #TODO
      # end

      def test_scrape
        #this doesn't test stashing.
        start_test_server()

        headlines = ["Webinar: How to Use Prescriber Checkup to Power Your Reporting", 
                     "",
                     "A Prosecutor, a Wrongful Conviction and a Question of Justice",
                     "Six Facts Lost in the IRS Scandal"]

        propubscraper = Upton::Scraper.new("http://127.0.0.1:9876/propublica.html", "section#river section h1 a", :css)
        propubscraper.debug = true
        propubscraper.verbose = true

        heds = propubscraper.scrape do |article_str|
          doc = Nokogiri::HTML(article_str)
          hed = doc.css('h1.article-title').text
        end
        assert_equal(heds, headlines)
        FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
      end

      def test_encodings
        skip "should test getting pages, switching their encoding to UTF-8, saving them as UTF-8, reading them as UTF-8"
      end

      def test_stashing
        skip "should test stashing, make sure we never send too many requests"
      end

      def test_scrape_list
        #this doesn't test stashing.
        #TODO: needs a website that has links to a multi-page list (or table)
        start_test_server()

        most_commented_heds = [["Six Facts Lost in the IRS Scandal", 
                            "How the IRS’s Nonprofit Division Got So Dysfunctional", 
                            "Sound, Fury and the IRS Mess", 
                            "The Most Important #Muckreads on Rape in the Military", 
                            "Congressmen to Hagel: Where Are the Missing War Records?", 
                            "As Need for New Flood Maps Rises, Congress and Obama Cut Funding", 
                            "A Prosecutor, a Wrongful Conviction and a Question of Justice", 
                            "A Prolonged Stay: The Reasons Behind the Slow Pace of Executions", 
                            "The Story Behind Our Hospital Interactive",
                            "irs-test-charts-for-embedding"]]

        propubscraper = Upton::Scraper.new(["http://127.0.0.1:9876/propublica.html"])
        propubscraper.debug = true
        propubscraper.verbose = true
        list = propubscraper.scrape(&Upton::Utils.list("#jamb.wNarrow #most-commented li a", :css))

        assert_equal(list, most_commented_heds)
        FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
      end

      def test_scrape_table
        #this doesn't test stashing.
        start_test_server()

        east_timor_prime_ministers = [[ 
                                        ["#", "Portrait", "Name(Birth–Death)", "Term of Office", "Party", 
                                          "1", "2", "3", "4",],
                                        [],
                                        ["", "Mari Alkatiri(b. 1949)", "20 May 2002", "26 June 2006[1]", "FRETILIN"],
                                        ["", "José Ramos-Horta(b. 1949)", "26 June 2006", "19 May 2007", "Independent"],
                                        ["", "Estanislau da Silva(b. 1952)", "19 May 2007", "8 August 2007", "FRETILIN"],
                                        ["", "Xanana Gusmão(b. 1946)", "8 August 2007", "Incumbent", "CNRT"],
                                      ]]

        propubscraper = Upton::Scraper.new(["http://127.0.0.1:9876/easttimor.html"])
        propubscraper.debug = true
        propubscraper.verbose = true
        table = propubscraper.scrape(&Upton::Utils.table('//table[contains(concat(" ", normalize-space(@class), " "), " wikitable ")][2]'))
        assert_equal(table, east_timor_prime_ministers)
        FileUtils.rm_r("test_stashes") if Dir.exists?("test_stashes")
      end



      private
      def start_test_server
        @server_thread = Thread.new do
          Rack::Handler::Thin.run Upton::Test::Server.new, :Port => 9876
        end
        sleep(1) # wait a sec for the server to be booted
      end
    end



    # via http://stackoverflow.com/questions/10166611/launching-a-web-server-inside-ruby-tests
    class Server
      def call(env)
        @root = File.expand_path(File.dirname(__FILE__))
        path = Rack::Utils.unescape(env['PATH_INFO'])
        path += 'index.html' if path == '/'
        file = File.join(@root, "data", path)

        params = Rack::Utils.parse_nested_query(env['QUERY_STRING'])

        if File.exists?(file)
          [ 200, {"Content-Type" => "text/html; charset=utf-8"}, File.read(file) ]
        else
          [ 404, {'Content-Type' => 'text/plain'}, 'file not found' ]
        end
      end
    end
  end
end

spec/spec_helper.rb
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# Require this file using `require "spec_helper"` to ensure that it is only
# loaded once.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration

require "webmock/rspec"
RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
  config.filter_run :focus

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = 'random'
  WebMock.disable_net_connect!(:allow_localhost => true)
end


lib/upton.rb
# encoding: UTF-8

require_relative 'upton/scraper'
require_relative 'upton/utils'
require_relative 'upton/version'
require_relative 'upton/downloader'
require_relative 'upton/scraper'

##
# This module contains a scraper called Upton
##
module Upton
  ##
  # *Upton* is a framework for easy web-scraping with a useful debug mode
  # that doesn't hammer your target's servers. It does the repetitive parts of
  # writing scrapers, so you only have to write the unique parts for each site.
  #
  # Upton operates on the theory that, for most scraping projects, you need to
  # scrape two types of pages:
  #
  # 1. Index pages, which list instance pages. For example, a job search
  #     site's search page or a newspaper's homepage.
  # 2. Instance pages, which represent the goal of your scraping, e.g.
  #     job listings or news articles.
  ##

end


lib/upton/utils.rb
# encoding: UTF-8

##
# This module contains a collection of helpers for Upton
##
module Upton

  ##
  # This class contains a collection of helpers for Upton
  #
  # Each method returns a Proc that (with an & ) can be used as the final
  # argument to Upton's `scrape` and `scrape_to_csv`
  ##
  module Utils

    ##
    # Scrapes an HTML <table> element into an Array of Arrays. The header, if
    # present, is returned as the first row.
    ##
    def self.table(table_selector, deprecated=nil)
      return Proc.new do |html|
        output = []
        headers = html.search(table_selector).css("th").map &:text
        output << headers

        table = html.search(table_selector).css("tr").each{|tr| output << tr.css("td").map(&:text) }
        output
      end
    end

    ##
    # Scrapes any set of HTML elements into an Array. 
    ##
    def self.list(list_selector, deprecated=nil)
      return Proc.new do |html|
        html.search(list_selector).map{|list_element| list_element.text }
      end
    end

  end
end

lib/upton/downloader.rb
require "fileutils"
require "open-uri"
require "tmpdir"
require "restclient"
require_relative "./version"

module Upton

  # This class is used internally to download and cache the webpages
  # that are requested.
  #
  # By default, the cache location is the output of `Dir.tmpdir`/upton.
  # The Dir.tmpdir returns the temporary directory of the operating system.
  # By default, the stashed files have a non-human-readable md5-based filename.
  # If `readable_stash_filenames` is true, they will have human-readable names.
  class Downloader

    MAX_FILENAME_LENGTH = 130 #for unixes, win xp+
    EMPTY_STRING = ''

    attr_reader :uri, :cache_location, :verbose
    def initialize(uri, options = {})
      @uri = uri
      @options = options
      @cache = options.fetch(:cache) { true }
      @cache_location = File.absolute_path(options[:cache_location] || "#{Dir.tmpdir}/upton")
      @verbose = options[:verbose] || false
      @readable_stash_filenames = options[:readable_filenames] || false
      initialize_cache!
    end

    def get
      if cache_enabled?
        puts "Stashing enabled. Will try reading #{uri} data from cache." if @verbose
        download_from_cache!
      else
        puts "Stashing disabled. Will download from the internet." if @verbose
        from_resource = true
        resp = download_from_resource!
        {:resp => resp, :from_resource => from_resource }
      end
    end

    private

    def make_request_for_resource!
      RestClient.get(uri)
    end

    def download_from_resource!
      begin
        puts "Downloading from #{uri}" if @verbose
        resp = make_request_for_resource!
        puts "Downloaded #{uri}" if @verbose
      rescue RestClient::ResourceNotFound
        puts "404 error, skipping: #{uri}" if @verbose
      rescue RestClient::InternalServerError
        puts "500 Error, skipping: #{uri}" if @verbose
      rescue URI::InvalidURIError
        puts "Invalid URI: #{uri}" if @verbose
      rescue RestClient::RequestTimeout
        puts "Timeout: #{uri}" if @verbose
        retry
      end
      resp ||= EMPTY_STRING
    end

    def download_from_cache!
      resp = if cached_file_exists?
              puts "Cache of #{uri} available" if @verbose
              from_resource = false
              open(cached_file).read
            else
              if @verbose
                if @readable_stash_filenames
                  puts "Cache of #{uri} unavailable at #{filename_from_uri}. Will download from the internet"
                else
                  puts "Cache of #{uri} unavailable. Will download from the internet"
                end
              end
              from_resource = true
              download_from_resource!
            end
      unless cached_file_exists?
        if @verbose
          if @readable_stash_filenames
            puts "Writing #{uri} data to the cache at #{cached_file}"
          else
            puts "Writing #{uri} data to the cache"
          end
        end
        commented_resp = add_comment(resp)
        open(cached_file, 'w'){|f| f << commented_resp}
      end
      {:resp => resp, :from_resource => from_resource }
    end

    def add_comment(resp)
      # n = Nokogiri::HTML("<html></html>")
      # c = Nokogiri::XML::Comment.new(n, "asdfasdf")
      # n.root.add_child(c)
      # <!----Retrieved by Upton from http://www.somesite.com on January 15 at 4:28 p.m.-->
      msg = "Stashed file retrieved by Upton #{Upton::VERSION} from #{@uri} at #{Time.now}"
      resp_html = Nokogiri::HTML(resp)
      comment = Nokogiri::XML::Comment.new(resp_html, msg)
      if resp_html.root.nil?
        return resp
      elsif resp_html.root.children.empty?
        resp_html.root.add_child(comment)
      else
        resp_html.root.children.before(comment)
      end
      resp_html.to_html
    end

    def cache_enabled?
      !!@cache
    end

    def filename_from_uri
      @readable_stash_filenames ? readable_filename_from_uri : hashed_filename_from_uri
    end

    def hashed_filename_from_uri
      Digest::MD5.hexdigest(uri)
    end

    def readable_filename_from_uri
      html = "html"
      clean_url_max_length = MAX_FILENAME_LENGTH - html.length - cache_location.size
      clean_url = uri.gsub(/[^A-Za-z0-9\-_]/, "")[0...clean_url_max_length]
      "#{clean_url}.#{html}"
    end

    def cached_file
      "#{cache_location}/#{filename_from_uri}"
    end

    def cached_file_exists?
      File.exists?(cached_file)
    end

    def initialize_cache!
      unless Dir.exists?(cache_location)
        Dir.mkdir(cache_location)
        FileUtils.chmod 0700, cache_location
      end
    end

  end
end
